mesh/mesh.go

```diff
@@ -545,6 +546,11 @@ func (msh *Mesh) revertState(logger log.Log, revertTo types.LayerID) error {
 		return fmt.Errorf("revert state to layer %v: %w", revertTo, err)
 	}
 	logger.With().Info("successfully reverted state", log.Stringer("state_root", root))
+
+	if err = rewards.RevertToLayer(msh.db, revertTo); err != nil {
```

Comment:
this revert needs to be atomic with accounts revert. also rewards should be saved in the vm as well, so that accounts changes are atomic with rewards

Reply:
2 questions:
- why should we save rewards in vm? when the layers are reverted, they account balance were just reverted. why should we treat rewards differently inside vm?
- the mesh data and vm are are updated separately in the codebase right now. we have not discussed any effort to keep their operation atomic. i don't see how we can do this easily either.

Reply:
> why should we save rewards in vm? when the layers are reverted, they account balance were just reverted. why should we treat rewards differently inside vm?

so lets say it is not atomic, node failed to revert rewards. now api will provide rewards that are inconsistent with the actual balance (or vice versa if they are updated in a different order)

> the mesh data and vm are are updated separately in the codebase right now. we have not discussed any effort to keep their operation atomic. i don't see how we can do this easily either.

so first of all current codebase has consistency bugs, just for now we ignore them
also what is a mesh data?
- rewards
- applied layer
- transaction status (failure/success and for genesis we will add fake-receipts)
- anything else?

there are 2 options:
- change this things atomically in a single tx, this is viable since all that data is managed by vm. this approach is the simplest code wise and fail safe
- manually restore consistency of the data. so for example we need to answer how we handle consistency of the balances<>transactions and balances<>rewards, and transactions<>rewards in the event of the crash. in practice manual solution almost certainly will have tricky bugs. maybe always do vm revert if applied layer wasn't updated, but this is not a fail-safe approach

Reply:
> > why should we save rewards in vm? when the layers are reverted, they account balance were just reverted. why should we treat rewards differently inside vm?
> 
> so lets say it is not atomic, node failed to revert rewards. now api will provide rewards that are inconsistent with the actual balance (or vice versa if they are updated in a different order)
> 
> > the mesh data and vm are are updated separately in the codebase right now. we have not discussed any effort to keep their operation atomic. i don't see how we can do this easily either.
> 
> so first of all current codebase has consistency bugs, just for now we ignore them also what is a mesh data?
> 
> * rewards
> * applied layer
> * transaction status (failure/success and for genesis we will add fake-receipts)
> * anything else?
> 
> there are 2 options:
> 
> * change this things atomically in a single tx, this is viable since all that data is managed by vm. this approach is the simplest code wise and fail safe
> * manually restore consistency of the data. so for example we need to answer how we handle consistency of the balances<>transactions and balances<>rewards, and transactions<>rewards in the event of the crash. in practice manual solution almost certainly will have tricky bugs. maybe always do vm revert if applied layer wasn't updated, but this is not a fail-safe approach

the 1st is the obvious choice here. but my reservation is that accounts data are in sqlite only for genesis. if we make rewards and txs applied atomically, does that mean when we have a VM and account state trie, we will need to figure this out again?

if memory serves, i think geth simply apply rewards as internal transfers and record the block coinbase in the block header. so maybe rewards and accounts tables can be removed together once state is properly implemented.

Reply:
> the 1st is the obvious choice here. but my reservation is that accounts data are in sqlite only for genesis. if we make rewards and txs applied atomically, does that mean when we have a VM and account state trie, we will need to figure this out again?

there is an option to keep state in the node and provide it to the svm over some interface. it can be cgo api, or share a C reference to sqlite object 

> if. memory serves, i think geth simply apply rewards as internal transfers and record the block coinbase in the block header. so maybe rewards and accounts tables can be removed together once state is properly implemented.

state for accounts can remain in the flat storage (sql table). there is nothing improper about it. commitment (e.g. inner nodes of the trie) can be stored separately. basically we will follow erigon design, i described it some time ago. 

specifically accounts table doesnt have to be removed. it can be extended with accounts_inner_nodes table that will store inner nodes of the commitment tree. 

if you want rewards can be removed even now, but we will also have to drop (rework?) API for them

Reply:
for now i changed to atomic update with accounts data.